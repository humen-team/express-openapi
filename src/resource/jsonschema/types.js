/* Cast a string value to a JSON Schema primitive type.
 */
export default function castType(value, type) {
    if (type === 'integer' || type === 'long') {
        return parseInt(value, 10);
    }
    if (type === 'float' || type === 'double') {
        return parseFloat(value);
    }
    if (type === 'boolean') {
        return value.toLowerCase() === 'true' || value === '1';
    }
    if (type === 'date' || type === 'dateTime') {
        return new Date(value);
    }
    if (type === 'byte') {
        return Buffer.from(value, 'base64');
    }
    if (type === 'binary') {
        return Buffer.from(value, 'binary');
    }
    return value;
}
