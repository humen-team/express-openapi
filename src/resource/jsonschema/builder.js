import { isPlainObject, mapValues, merge } from 'lodash';

import { OPENAPI_2_0 } from '../../constants';

/* Does this schema have a nullable type?
 */
export function isNullable(object) {
    const { type } = object;
    return type === 'null' || (type?.includes && type.includes(null));
}

/* Build valid OpenAPI from JSON Schema.
 */
export default function buildOpenAPI(object, openapiVersion, path = '') {
    // recurse into lists
    if (Array.isArray(object)) {
        return object.map(
            (item) => buildOpenAPI(item, openapiVersion, `${path}[]`),
        );
    }

    const nullable = openapiVersion === OPENAPI_2_0 ? 'x-nullable' : 'nullable';

    // recurse into objects
    if (isPlainObject(object)) {
        return merge(
            // special case #1: nullability can be encoded
            object.type ? {
                [nullable]: isNullable(object),
            } : {},
            mapValues(
                object,
                (value, key) => {
                    // special case #2: types cannot be lists or contain `null`
                    if (key === 'type' && !isPlainObject(value)) {
                        return value?.filter
                            ? value.filter((item) => item !== null)[0]
                            : value;
                    }
                    // special case #3: items is a reference to another part of the document
                    if (key === '$ref') {
                        return value.build(openapiVersion);
                    }
                    return buildOpenAPI(value, openapiVersion, `${path}.${key}`);
                },
            ),
        );
    }

    // return primitives
    return object;
}
