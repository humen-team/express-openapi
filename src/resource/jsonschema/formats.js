import { validate } from 'uuid';

export function uuidFormat(input) {
    return input === null || validate(input);
}

export function userIdFormat(input) {
    return input === 'me' || uuidFormat(input);
}
