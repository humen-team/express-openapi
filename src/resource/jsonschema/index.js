import JSONSchemaResource from './resource';

export default JSONSchemaResource;
