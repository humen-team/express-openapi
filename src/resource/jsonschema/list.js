import JSONSchemaResource from './resource';

export default class JSONSchemaListResource extends JSONSchemaResource {
    constructor(item) {
        super(JSONSchemaListResource.createSchema(item), item.config);
        this.item = item;
    }

    static createSchema(item) {
        const { config } = item;
        const properties = {
            items: {
                type: 'array',
                items: item.toRef(),
            },
        };
        const required = [
            'items',
        ];

        // enable response to report on current offset/limit paging
        if (config.pagingStrategy.hasOffsetLimit) {
            properties.limit = {
                type: 'integer',
            };
            properties.offset = {
                type: 'integer',
            };
        }
        // enable response to report on total collection count
        if (config.pagingStrategy.hasCount) {
            properties.count = {
                type: 'integer',
            };
            required.push('count');
        }

        return {
            id: config.namingStrategy.toListName(item.id),
            type: 'object',
            properties,
            required,
        };
    }
}
