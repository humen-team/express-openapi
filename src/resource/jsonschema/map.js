import { get, mapValues } from 'lodash';

/* Map a function over an object and a schema.
 */
export default function mapSchema(value, schema, registry, func, path = '') {
    if (!schema) {
        return func(value, null);
    }

    const { $ref, type } = schema;

    if ($ref) {
        // XXX resolved Reference
        return mapSchema(
            value,
            registry[$ref],
            registry,
            func,
            `${path}.$ref`,
        );
    }

    if (type === 'array') {
        return value.map(
            (child) => mapSchema(
                child,
                get(schema, 'items'),
                registry,
                func,
                `${path}.items`,
            ),
        );
    }

    if (type === 'object') {
        return mapValues(
            value,
            (child, key) => mapSchema(
                child,
                get(schema, `properties.${key}`),
                registry,
                func,
                `${path}.properties.${key}`,
            ),
        );
    }

    return func(value, schema);
}
