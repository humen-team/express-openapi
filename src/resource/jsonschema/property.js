import { merge, omit } from 'lodash';

import Reference from '../reference';
import { OPENAPI_2_0 } from '../../constants';

/* Does this schema have a nullable type?
 */
export function isNullable(schema) {
    const { type } = schema;
    return type === 'null' || (type?.includes && type.includes(null));
}

/* Build a schema property as valid OpenAPI.
 */
export function buildProperty(schema, openapiVersion) {
    const { items, type } = schema;
    const nullable = openapiVersion === OPENAPI_2_0 ? 'x-nullable' : 'nullable';

    // special case #1: nullability is encoded differently
    const converted = {
        [nullable]: isNullable(schema),
    };

    // special case #2: types cannot be lists or contain `null`
    converted.type = type?.filter
        ? type.filter((item) => item !== null)[0]
        : type;

    // special case #3: items is a reference to another part of the document
    if (items instanceof Reference) {
        converted.items = items.build(openapiVersion);
    } else {
        converted.items = items;
    }

    return merge(
        omit(schema, ['items', 'type']),
        converted,
    );
}
