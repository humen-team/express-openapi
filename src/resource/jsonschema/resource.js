import { difference, mapValues, omit } from 'lodash';
import { validate } from 'jsonschema';

import { UnprocessableEntity } from '../../errors';
import Reference from '../reference';
import Resource from '../resource';
import { buildProperty } from './property';
import castType from './types';

/* Defines an OpenAPI compatible resource using JSONSchema.
 */
export default class JSONSchemaResource extends Resource {
    constructor(schema, config) {
        super(config);
        this.schema = schema;
    }

    get id() {
        return this.config.namingStrategy.toName(this.schema.id);
    }

    get properties() {
        return this.schema.properties;
    }

    get required() {
        return this.schema.required;
    }

    get keys() {
        return Object.keys(this.schema.properties || {});
    }

    /* Convert this resource to an OpenAPI compatible definition.
     *
     * While OpenAPI leverages JSONSchema, it isn't 100% compatible; some conversion is required.
     */
    build(openapiVersion) {
        const { properties, required, type } = this.schema;
        return {
            // omit the `id` field
            properties: mapValues(
                properties,
                (value) => buildProperty(value, openapiVersion),
            ),
            required,
            type,
        };
    }

    cast(data) {
        return mapValues(
            data,
            (value, key) => castType(value, this.schema.properties[key].type),
        );
    }

    toList() {
        // avoid circular dependency by deferring import
        const JSONSchemaResourceList = require('./list').default; // eslint-disable-line global-require
        return new JSONSchemaResourceList(this);
    }

    toRef() {
        return new Reference(this.id);
    }

    /* Validate data against this JSONSchema schema.
     */
    validate(data) {
        const result = validate(data, this.schema);
        if (result.errors.length) {
            // NB: we could report on mutiple errors
            const { instance, message } = result.errors[0];
            throw new UnprocessableEntity({
                message: `${instance} ${message}`,
            });
        }
        return data;
    }

    /* Remove some properties from a resource.
     */
    without(id, ignore) {
        return new JSONSchemaResource({
            id,
            type: 'object',
            properties: omit(this.properties, ignore),
            required: difference(this.required, ignore),
        });
    }
}
