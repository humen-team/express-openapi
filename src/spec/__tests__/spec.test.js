import { readFileSync } from 'fs';
import { Validator } from 'jsonschema';

import Spec from '../spec';

describe('Spec', () => {
    const openapi2 = JSON.parse(readFileSync('schemas/openapi/2.0/schema.json'));
    const openapi3 = JSON.parse(readFileSync('schemas/openapi/3.0/schema.json'));
    const jsonSchema = JSON.parse(readFileSync('schemas/json-schema/draft-04/schema.json'));
    const validator = new Validator();
    validator.addSchema(jsonSchema);

    describe('build', () => {
        it('encodes a valid swagger 2.0 spec with no operations', () => {
            const spec = new Spec().build('2.0');
            expect(spec).toMatchSnapshot();
            validator.validate(spec, openapi2, { throwError: true });
        });

        it('encodes a valid swagger 3.0 spec with no operations', () => {
            const spec = new Spec().build('3.0.0');
            expect(spec).toMatchSnapshot();
            validator.validate(spec, openapi3, { throwError: true });
        });
    });
});
