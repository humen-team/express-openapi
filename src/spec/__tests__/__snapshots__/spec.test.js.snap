// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Spec build encodes a valid swagger 2.0 spec with no operations 1`] = `
Object {
  "basePath": "/",
  "definitions": Object {},
  "info": Object {
    "description": "REST API",
    "title": "REST API",
    "version": "0.1.0",
  },
  "paths": Object {},
  "schemes": Array [
    "http",
  ],
  "swagger": "2.0",
}
`;

exports[`Spec build encodes a valid swagger 3.0 spec with no operations 1`] = `
Object {
  "components": Object {
    "schemas": Object {},
  },
  "info": Object {
    "description": "REST API",
    "title": "REST API",
    "version": "0.1.0",
  },
  "openapi": "3.0.0",
  "paths": Object {},
  "servers": Array [
    Object {
      "url": "/",
    },
  ],
}
`;
